TypeScript support (draft)

benjamincburns opened the pull request:
Adds support for TypeScript code execution by running TypeScript code via any supported installed TypeScript Jupyter Kernel.

Also handles type checking prior to execution using `deno check`.

pawamoy commented on src/markdown_exec/__init__.py:
```diff
@@ -24,10 +24,13 @@
 
 __all__: list[str] = ["formatter", "validator"]
 
-MARKDOWN_EXEC_AUTO = [lang.strip() for lang in os.getenv("MARKDOWN_EXEC_AUTO", "").split(",")]
+MARKDOWN_EXEC_AUTO = [
+    lang.strip() for lang in os.getenv("MARKDOWN_EXEC_AUTO", "").split(",")
+]
```
style: Early note to recommend disabling auto-formatting in your IDE/editor, or have it respect our config in `config/ruff.toml` :slightly_smiling_face:

benjamincburns replied:
will do! good to get ahead of this, but also I haven't thought about those final "polishing" steps at all for this yet 😄

pawamoy commented on src/markdown_exec/formatters/base.py:
```diff
@@ -141,13 +143,19 @@ def base_format(
     except ExecutionError as error:
         identifier = id or extra.get("title", "")
         identifier = identifier and f"'{identifier}' "
-        exit_message = "errors" if error.returncode is None else f"unexpected code {error.returncode}"
+        exit_message = (
+            "errors"
+            if error.returncode is None
+            else f"unexpected code {error.returncode}"
+        )
         log_message = (
-            f"Execution of {language} code block {identifier}exited with {exit_message}\n\n"
+            f"Execution of {language} code block {identifier} exited with {exit_message}\n\n"
```
bug: `identifier` already contains the trailing space.

pawamoy commented on src/markdown_exec/formatters/typescript.py:
```diff
@@ -0,0 +1,197 @@
+                )
+                return output
+
+        except BrokenPipeError:
+            raise ExecutionError("Deno process died unexpectedly")
+    finally:
+        _exec_lock.release()
```
lint: I'd recommend avoiding such large `try` blocks. Maybe split out the inner logic into separate functions. Ideally the `try` blocks have only one line.

benjamincburns replied:
yeah, this was a result of just (ahem) trying things to understand where my race condition is coming from - I was originally using the context manager protocol on `Lock` for this (so it was a `with` block rather than a `try`/`finally`), but at one stage I was finding that things were processing out of order so I decided to make the lock/unlock more explicit.

worst case if I need to keep the try/finally I'll break this down into internal functions with more readable names and then call them from the `try` block.

pawamoy reviewed (commented):
Nice!

> No process lifetime management 

I suppose it would be better to spawn the subprocess when the typescript formatter is called the first time indeed. Could use a `@functools.cached` decorated function that accepts no input and returns None for this :slightly_smiling_face: 

> Ideally the lifetime of this process would be tied to the rendering lifetime of a single markdown document.
>
> ---
>
> No session lifetime management
> Because there's one process used for all documents rendered, if two documents use the same session name, they'll share context.

Actually I think it is good that sessions are shared across documents. That's how the Python formatter works, and it's a useful feature. Users are already relying on it :slightly_smiling_face: 

So, Our MkDocs plugin could terminate it in a post build hook. Outside of MkDocs (when Markdown-Exec is just used as a Markdown extension), I'm not sure when we should terminate it :thinking: Maybe we don't and let the OS terminate it when the parent Python process finishes :shrug:

benjamincburns commented:
> That's how the Python formatter works, and it's a useful feature. Users are already relying on it 🙂

That's good to know - I think for our use this could potentially cause some confusion though. @eyurtsev - FYI

> So, Our MkDocs plugin could terminate it in a post build hook. Outside of MkDocs (when Markdown-Exec is just used as a Markdown extension), I'm not sure when we should terminate it 🤔 Maybe we don't and let the OS terminate it when the parent Python process finishes 🤷

Yeah, that all aligns with what I was thinking. I'm not super familiar with MkDocs, but I was going to go looking for a hook like that for termination. If there's something like that that we can optionally trigger after each doc, that would also make the cross-document sessionkeeping behavior configurable.

pawamoy commented:
> If there's something like that that we can optionally trigger after each doc, that would also make the cross-document sessionkeeping behavior configurable.

Good idea. I'm OK to add a configuration option to choose between scoping sessions to documents, or have it global for the whole build. We can then use the `on_page_content` hook to terminate the subprocess after each Markdown conversion, or the `on_post_build` hook to terminate the subprocess after the whole site is built. Then each time we want to execute a TypeScript code block, we re-spawn the subprocess if needed (global variable that oscillates between the process instance and none).

pawamoy commented:
Wow that's a lot of code :joy: Not against merging it, but I'll have to ask if you will be available to help maintain it in the future if we get issues about it :smile: Also when you have the time, I'd love to learn why you chose to go the Jupyter kernel way. I guess one immediate benefit is that we could use that to support these languages: https://github.com/jupyter/jupyter/wiki/Jupyter-kernels.

benjamincburns commented:
> Wow that's a lot of code 😂 Not against merging it, but I'll have to ask if you will be available to help maintain it in the future if we get issues about it 😄

Yeah, it got a bit bigger than I first expected 😅 Re: maintaining it, I'm not sure exactly what I can commit to, to be honest. I'm submitting this PR so we can use it in the docs pipeline for LangGraph. Assuming that all goes well, I think we'd have a vested interest in helping to maintain it when it comes to things that either break or would substantially add value to our docs. Outside of that, I'm not as sure. If you'd prefer that we consume this from a fork that credits this project as the original, I can totally understand.
 
> Also when you have the time, I'd love to learn why you chose to go the Jupyter kernel way. I guess one immediate benefit is that we could use that to support these languages: https://github.com/jupyter/jupyter/wiki/Jupyter-kernels.

It was fairly straightforward to get TypeScript type checking and JavaScript code execution working with very minimal additional dependencies (namely deno, which ships as a single statically-linked binary). Because of the need to support sessions, I can't just execute the code block as a script. I need some process that executes the code in blocks, maintaining the context as I go. Also I wanted to avoid having to figure out how to parse the process-level streams myself, as the runtime can sometimes inject its own diagnostics output.

That lead me to my initial approach, where I spawned a single `deno` execution process running a custom execution script, and communicated with it via a simple JSON-based request/response protocol. In doing this, I naïvely thought that because Deno supports TypeScript execution out of the box, it'd let me `eval` typescript code (or more correctly, it'd let me run code in a "sandboxed" context using e.g. [`vm.SourceTextModule`](https://nodejs.org/api/vm.html#class-vmsourcetextmodule).

Unfortunately deno doesn't automatically transform eval'd code, so I would've needed an explicit code transform step to get it working that way. That would've meant introducing JS dependencies, which would've either required the consumer to install those themselves, or a more complex node module resolution pipeline that'd pull in those deps from the `markdown-exec` directory structure before switching to the user's directory to resolve dependencies from within their project. Also, If you clicked through to the docs, you may have also noticed that `SourceTextModule` is still an experimental API (even though some very popular libs like `vitest` and `jest` use it).

I could've done all that, but I've actually tried before on a different project and it was rather painful. Then I realized - there's a project that does literally _all_ of the things I need already - Jupyter. Better still, I can install a JS/TS kernel in a single command with no additional JS deps required thanks to `deno jupyter --install`. I also figured that if I added a generic Jupyter formatter path, it'd make it trivially easy to support execution of many other languages (as you already spotted), and the captured outputs would be more predictable to people such as myself who've used Jupyter but don't like the friction of editing notebook files.

It's a lot of lines now, but it'll amortize well if you want to open the floodgates and support any language that ships a Jupyter kernel.

benjamincburns commented:
Regarding the test failure - I'll look into it. It appears to be due to jupyter_client fetching the paths for jupyter config files using the old path scheme. If I set the env var as suggested in the warning it doesn't pick up the config from the `deno` install, however (at least not locally). It could potentially be solvable with a simple tweak, or I may need to pin to an older version of `jupyter_core` until deno supports installing to the new path scheme - not sure.

benjamincburns commented:
Ah, and FWIW, I think the history feature of jupyter kernels _may_ make #4 easier to achieve. You'd need to switch the other languages to route through jupyter kernels, but I think it'd be fairly easily done - and you'd get more uniform feature support (e.g. sessions) everywhere.

benjamincburns commented on .github/workflows/ci.yml:
```diff
@@ -46,6 +50,21 @@ jobs:
+    - name: Install Deno Jupyter Kernel
+      run: deno jupyter --install
+
+    - name: Fix deno path on macOS
+      if: runner.os == 'macOS'
+      run: |
+        export JUPYTER_PLATFORM_DIRS=0
+        ORIG_DENO_INSTALL_PATH="$(dirname "$(jupyter kernelspec list | grep deno | sed 's/[^\/]*//')")"
+        mkdir -p "$HOME/Library/Application Support/Jupyter"
+        cp -r "$ORIG_DENO_INSTALL_PATH" "$HOME/Library/Application Support/Jupyter/kernels/deno"
+        export JUPYTER_PLATFORM_DIRS=0
+
+        # fail if jupyter still can't find the kernel config
+        jupyter kernelspec list | grep -q deno
```
This is unfortunately necessary due to denoland/deno#27984

benjamincburns commented:
I think I'll probably need to ignore tests for this on 3.14 due to https://github.com/jupyter/jupyter_client/issues/1053

benjamincburns commented:
Also I'd love to know what's unique about that 3.11, highest config that keeps failing - I can't reproduce it locally (tried on ubuntu and macOS) 🤔

pawamoy commented:
>  think we'd have a vested interest in helping to maintain it when it comes to things that either break or would substantially add value to our docs.

Thanks, I'm not asking anything more than being available to have a look if/when things break for this particular part of the project :slightly_smiling_face: With time I should be able to "own" that part too, so it's not a commitment for life either.

---

Thank you so much for the explanation, this was super clear, and I do see the advantage of using Jupyter kernels since they bring the possibility to support many, many more languages in a unified manner (with support for sessions!).

---

> make check fails at the breaking change check because I don't have all the tags pulled.

Yep don't worry about this check, it's in "allowed failure" mode (both locally and in CI).

---

Regarding solving #4 with the history feature of Jupyter kernels, I probably don't know enough of Jupyter kernels to say anything about that. My feeling is that I'd be much more comfortable just relying on the output of the system, regardless of the system itself producing this output.

---

RE: tests on 3.14: yep, don't bother with those either, they're allowed to fail too. Many packages don't work yet on Python 3.14.

---

> Also I'd love to know what's unique about that 3.11, highest config that keeps failing 

Hmm I've seen similar issues in the past, let me look into it.

pawamoy commented:
I'm getting the same error on every Python version. The difference between versions in CI might be due to caching :thinking:? Still, I don't understand why code gets highlighted since our `md` fixture doesn't add any highlighting extension.

pawamoy commented:
Hmm OK locally my errors are due to `deno` not being installed.

pawamoy commented:
One issue with this is that the `FileNotFoundError` wasn't caught, and later silenced by superfences. I think we should catch any exception in either `_run_jupyter` or `_run_typescript`. The former only catches `queue.empty` and checks for errors to raise `ExecutionError`: it should raise `ExecutionError` for any other (currently uncaught) exception.

pawamoy commented:
OK looks like one of `pymdownx.superfences` or `pymdownx.tabbed` *also* registers `pymdownx.highlight`. Maybe we could remove it manually.

pawamoy commented:
OK, superfences registers highlight in "disabled" mode, but when we mimic the Markdown instance we lose track of this mode. I'll push a bugfix.

pawamoy commented:
Well it actually makes sense that superfences enables highlighting. IMO our tests should expect execution errors to be highlighted. Not sure why the other CI jobs passed though. Try to use `"..." in html` instead of exact match in your tests maybe?

benjamincburns commented:
> IMO our tests should expect execution errors to be highlighted.

Yeah, that's why I wrote the `_expectError` utility fn in the `test_typescript` file. I was comparing against the non-highlighted format strings and of course that wasn't going so well.

As for why the highlighting is happening, it's because I followed how `ExecutionError` was used elsewhere and wrapped most error messages in a `code_block` call. I figured you were doing that to handle cases with error formats that would be parsed out by the code highlighter.

> I think we should catch any exception in either `_run_jupyter` or `_run_typescript`.

Ideally this should be done more centrally, I think? I'll look at the call stack and see if there isn't a more fitting place. If not, I'll add it where you suggest.

benjamincburns commented:
Also I should've said in my last message, but thanks again for the fast & thorough responses here! I have a goal to get to a go/no-go decision on whether this strategy will work for our docs by the end of today (it's currently the start of my day - I'm based in NZ), and I don't think I'd have been able to without your involvement.

And yep, good to know re: maintenance commitment. I'll do the few things I said above, and then switch over to actually trying to use this in a modified version of our docs pipeline. So far I've only tested it w/ the docs in this project. Provided everything goes well and we're able to see a clear path toward using this, I'll remove the draft status.

pawamoy commented:
> I followed how ExecutionError was used elsewhere and wrapped most error messages in a code_block call. I figured you were doing that to handle cases with error formats that would be parsed out by the code highlighter.

The simple reason why we raise execution errors with code blocks is that these code blocks are then displayed in the rendered docs :slightly_smiling_face: It was initially our only way of displaying errors (before we also added more verbose logging).

> Ideally this should be done more centrally, I think?

Right, maybe the base formatter is a good candidate (though not all formatters use it IIRC).

> thanks again for the fast & thorough responses

You're welcome! Sounds good, I'll let you evaluate the solution on your side at your own pace :slightly_smiling_face:

benjamincburns commented:
Oh btw - if you are only passingly familiar w/ Jupyter and/or iPython and want to become very well-acquainted with how kernels work relatively quickly - the [messaging spec doc](https://jupyter-client.readthedocs.io/en/stable/messaging.html) for the `jupyter-python` package lays out kernel interaction really nicely. It's what informed the implementation in this PR. The Jupyter project as a whole is documented exceptionally well, and I think that's a big part of why it's so successful.

benjamincburns commented:
> One issue with this is that the FileNotFoundError wasn't caught, and later silenced by superfences. I think we should catch any exception in either _run_jupyter or _run_typescript. The former only catches queue.empty and checks for errors to raise ExecutionError: it should raise ExecutionError for any other (currently uncaught) exception.
> 
> For example I installed deno and then another exception wasn't caught and silenced: ValueError: No kernel found for language \'typescript\'. Available kernels:.

I've added more central error handling, and made the error handling for these two cases throw `ExecutionError`s that have more explicit/actionable advice.

benjamincburns commented:
In theory the tests all pass. In practice, there's something flaky about `deno` that is throwing random panics, and I'm really not sure what's causing it. I did have a successful test run on everything but 3.14 as of 1182cd2, but then when I added some decorators to skip my tests on 3.14 I got another flaky failure 😞.

pawamoy commented on .github/workflows/ci.yml:
```diff
@@ -46,6 +50,59 @@ jobs:
+        
+        export JUPYTER_PLATFORM_DIRS=1
+
+        # make the data directory for the new path structure
+        mkdir -p "$(jupyter --data-dir)/kernels"
+
+        cp -r "$ORIG_JUPYTER_KERNEL_PATH/deno" "$(jupyter --data-dir)/kernels"
+
+        # fail if jupyter still can't find the kernel config
+        jupyter kernelspec list | grep -q deno
+        
+        deactivate
+
+    - name: Fix deno path on Windows
+      if: runner.os == 'Windows'
+      shell: pwsh
+      run: |
+        .\.venv\Scripts\Activate.ps1
+        $env:JUPYTER_PLATFORM_DIRS = "0"
+
+        # ensure jupyter can find the kernel config using the old path structure
+        $kernelCheck = jupyter kernelspec list | Select-String -Pattern "deno"
+        if (-not $kernelCheck) { exit 1 }
+
+        $origKernelPath = Join-Path (jupyter --data-dir) "kernels"
+        
+        $env:JUPYTER_PLATFORM_DIRS = "1"
+
+        # make the data directory for the new path structure
+        $newKernelPath = Join-Path (jupyter --data-dir) "kernels"
+        New-Item -ItemType Directory -Force -Path $newKernelPath
+
+        Copy-Item -Path (Join-Path $origKernelPath "deno") -Destination $newKernelPath -Recurse -Force
+
+        # fail if jupyter still can't find the kernel config
+        $kernelCheck = jupyter kernelspec list | Select-String -Pattern "deno"
+        if (-not $kernelCheck) { exit 1 }
+        
+        deactivate
+
```
issue (ci perf): We're in the `quality` job, which only runs on ubuntu-latest, so I don't think we need these.

pawamoy commented on .github/workflows/ci.yml:
```diff
@@ -118,17 +175,73 @@ jobs:
+        source .venv/bin/activate
+        export JUPYTER_PLATFORM_DIRS=0
+
+        # ensure jupyter can find the kernel config using the old path structure
```
style: Please capitalize and punctuate (period at the end) every comment in the PR :pray:

pawamoy commented on .github/workflows/release.yml:
```diff
@@ -20,6 +23,10 @@ jobs:
         python-version: "3.12"
     - name: Setup uv
       uses: astral-sh/setup-uv@v3
+    - name: Set up Deno
+      uses: denoland/setup-deno@v2
+    - name: Install Deno Jupyter Kernel
+      run: deno jupyter --install
```
issue (ci perf): I don't think we need this to build the package and create a GitHub release (nor the env var right above).

pawamoy commented on README.md:
```diff
@@ -65,7 +66,8 @@ markdown_extensions:
       format: !!python/name:markdown_exec.formatter
     # ...one fence for each language we support:
-    # bash, console, md, markdown, py, python, pycon, sh, tree
+    # bash, console, md, markdown, py, python, pycon, sh, tree, typescript, ts,
+    # tscon
```
style: For consistency with same change above:

```suggestion
    # bash, console, md, markdown, py, python, pycon, sh, tree,
    # typescript, ts, tscon
```

pawamoy commented on docs/usage/typescript.md:
```diff
@@ -0,0 +1,53 @@
+
+### Output capturing
+
+Outputs are captured just as they are in jupyter notebooks.
```
style:

```suggestion
Output is captured like in Jupyter notebooks.
```

pawamoy commented on docs/usage/typescript.md:
```diff
@@ -0,0 +1,53 @@
+
+## TypeScript REPL console code
+
+Code blocks syntax-highlighted with the `tscon` identifier are also supported.
```
question: Is `tscon` a thing? Like, does TypeScript have an interactive console/REPL too, like Python? Or did you just implement `tscon` because we have it for Python?

pawamoy commented on mkdocs.yml:
```diff
@@ -23,6 +23,7 @@ nav:
   - usage/index.md
   - Python: usage/python.md
   - Pyodide: usage/pyodide.md
+  - TypeScript: usage/typescript.md
```
style: Please sort alphabetically :slightly_smiling_face:

pawamoy commented on pyproject.toml:
```diff
@@ -29,6 +29,7 @@ classifiers = [
     "Typing :: Typed",
 ]
 dependencies = [
+    "jupyter-client>=8.6.3",
```
issue: I believe we should make it optional (will need update to the installation/usage docs), as users who are not interested in TypeScript will not want to pull in jupyter-client and all its dependencies. This might change in the future if we use Jupyter kernels for all languages but that's not the case for now :slightly_smiling_face:

pawamoy commented on src/markdown_exec/__init__.py:
```diff
@@ -124,7 +130,10 @@ def formatter(
+    try:
+        return fmt(code=source, md=md, **options)  # type: ignore[operator]
+    except Exception as e:
+        raise ExecutionError(traceback.format_exc()) from e
```
issue: Since we're returning a Python traceback, this should be formatted as a Python code block for proper display in the rendered docs.

pawamoy commented on src/markdown_exec/formatters/jupyter.py:
```diff
@@ -0,0 +1,253 @@
+    """Get information about all available Jupyter kernels.
+
+    Returns:
+        A dictionary mapping kernel names to their information including:
+        - display_name: Human readable name
+        - language: The programming language
```
lint: Python-Markdown requires new lines before lists:

```suggestion
        A dictionary mapping kernel names to their information including:

        - display_name: Human readable name
        - language: The programming language
```

pawamoy commented on src/markdown_exec/formatters/jupyter.py:
```diff
@@ -0,0 +1,253 @@
+
+    Returns:
+        A dictionary mapping kernel names to their information including:
+        - display_name: Human readable name
```
style: Please wrap code words in backticks and add period at the end of sentences :pray: :

```suggestion
        - `display_name`: Human readable name.
```

pawamoy commented on src/markdown_exec/formatters/jupyter.py:
```diff
@@ -0,0 +1,253 @@
+            "language_info": spec["spec"].get("language_info", {}),
+        }
+        for name, spec in kernel_specs.items()
+    }
+
+_language_to_kernel: dict[str, str] = {}
+
+def _get_kernel_for_language(language: str) -> str:
```
style: Looks like this code is not Ruff-formatted, could you run `make format` :smile:?

pawamoy commented on src/markdown_exec/formatters/jupyter.py:
```diff
@@ -0,0 +1,253 @@
+    no_kernel_msg = f"No jupyter kernel found that supports language '{language}'."
+
+    if language == "typescript":
+        no_kernel_msg += "Please install the deno jupyter kernel by running `deno jupyter install`."
```
```suggestion
        no_kernel_msg += "Please install the Deno Jupyter kernel by running `deno jupyter install`."
```

pawamoy commented on src/markdown_exec/formatters/jupyter.py:
```diff
@@ -0,0 +1,253 @@
+    _kernel_managers.clear()
+
+# Because this is a language agnostic runner, it has two extra initial language
+# arguments. This makes it easy to use functools.partial to create a language
```
```suggestion
# arguments. This makes it easy to use `functools.partial` to create a language
```

pawamoy commented on src/markdown_exec/formatters/jupyter.py:
```diff
@@ -0,0 +1,253 @@
+                if msg_type == "stream":
+                    outputs.append(content["text"])
+                elif msg_type in ("execute_result", "display_data"):
+                    # TODO: these can contain other MIME types (including HTML) - do we want to support them?
```
answer: Markdown Exec supports Markdown and HTML outputs, so it sounds like a good idea to also support HTML output. I'm not sure what that implies with Jupyter though.

pawamoy commented on src/markdown_exec/formatters/jupyter.py:
```diff
@@ -0,0 +1,253 @@
+    execution_completed = False
+
+    while not execution_completed:
+        try:
+            # Check for shell messages (execution replies)
+            shell_msg = kc.get_shell_msg(timeout=0.1)
+            if shell_msg["parent_header"].get("msg_id") == msg_id:
+                execution_completed = True
+
+                if shell_msg["content"]["status"] == "error":
+                    if (
+                        shell_msg["content"]["traceback"] is not None
+                        and len(shell_msg["content"]["traceback"]) > 0
+                    ):
+                        if shell_msg["content"]["traceback"][0] == "Stack trace:":
+                            error_outputs.extend(
+                                shell_msg["content"]["traceback"][1:],
+                            )
+                        else:
+                            error_outputs.extend(shell_msg["content"]["traceback"])
+                    break
+                # note: abort is deprecated, but some older kernels may still send it
+                # it doesn't send error info, however - so we'll need to populate a fake abort error instead
+                if shell_msg["content"]["status"] == "abort":
+                    error_outputs.append("AbortError: Execution aborted")
+                    break
+        except queue.Empty:
+            pass
```
lint: Same remark as before: what line exactly is susceptible of raising `queue.Empty`? Could we wrap only this line in `try`/`except`?

pawamoy commented on src/markdown_exec/formatters/jupyter.py:
```diff
@@ -0,0 +1,253 @@
+                # note: abort is deprecated, but some older kernels may still send it
+                # it doesn't send error info, however - so we'll need to populate a fake abort error instead
+                if shell_msg["content"]["status"] == "abort":
+                    error_outputs.append("AbortError: Execution aborted")
```
issue: It looks like `error_outputs` can contain plain text lines, which won't play well with them being wrapped as a code block when raising an execution error :thinking: Not sure how to best fix that though. Maybe keep this line intact, remove code block wrapping when raising the execution error, and add code block wrapping to the line above: 

```python
error_outputs.extend(code_block(language, shell_msg["content"]["traceback"]))
```

pawamoy commented on src/markdown_exec/formatters/jupyter.py:
```diff
@@ -0,0 +1,253 @@
+        except queue.Empty:
+            pass
+
+        try:
+            # nested while here so we consume all iopub messages for this execution before moving on.
+            # this lets us consume all pending messages after execution has completed before exiting the loop
+            while True:
+                # Check for iopub messages (output)
+                iopub_msg = kc.get_iopub_msg(timeout=0.1)
+                if iopub_msg["parent_header"].get("msg_id") != msg_id:
+                    continue
+
+                msg_type = iopub_msg["header"]["msg_type"]
+                content = iopub_msg["content"]
+
+                if msg_type == "stream":
+                    outputs.append(content["text"])
+                elif msg_type in ("execute_result", "display_data"):
+                    # TODO: these can contain other MIME types (including HTML) - do we want to support them?
+                    outputs.append(str(content["data"].get("text/plain", "")))
+                elif msg_type == "error":
+                    # ignore these because we already got it from the shell message
+                    pass
+
+        except queue.Empty:
```
lint: Again, could we wrap only the line susceptible of raising `queue.Empty`?

pawamoy commented on src/markdown_exec/formatters/jupyter.py:
```diff
@@ -0,0 +1,253 @@
+        **kwargs,
+    )
+
+# if _shutdown_kernels isn't called before the process exits, the windows build
+# of deno panics see https://github.com/denoland/deno/issues/28004 for more info
```
```suggestion
# if `_shutdown_kernel`s isn't called before the process exits, the Windows build
# of Deno panics, see https://github.com/denoland/deno/issues/28004 for more info
```

\+ capitalization and punctuation. Please use Markdown syntax (inline code, etc.) in every comment :pray:

pawamoy commented on src/markdown_exec/formatters/jupyter.py:
```diff
@@ -0,0 +1,253 @@
+
+# if _shutdown_kernels isn't called before the process exits, the windows build
+# of deno panics see https://github.com/denoland/deno/issues/28004 for more info
+atexit.register(_shutdown_kernels)
```
lint: I've had a bad experience with projects registering exit functions at the module-level. Please only do so if needed (if relevant part of the code is entered at runtime, not import time).

pawamoy commented on src/markdown_exec/formatters/typescript.py:
```diff
@@ -0,0 +1,129 @@
+
+    if not deno_path:
+        raise ExecutionError(
+            "Deno not found. Please install it from "
+            "https://deno.land/manual/getting_started/installation. Be sure to "
+            "also install the deno jupyter kernel by running `deno jupyter "
+            "install`.",
```
```suggestion
            "Deno not found. "
            "Please install it from https://deno.land/manual/getting_started/installation. "
            "Be sure to also install the Deno Jupyter kernel by running `deno jupyter install`.",
```

pawamoy commented on src/markdown_exec/mkdocs_plugin.py:
```diff
@@ -128,6 +129,8 @@ def on_post_build(self, *, config: MkDocsConfig) -> None:  # noqa: ARG002,D102
         else:
             os.environ["MKDOCS_CONFIG_DIR"] = self.mkdocs_config_dir
 
+        _shutdown_kernels()
```
issue: `on_post_build` runs after *each* build during live-reload, so every time the site is rebuilt/reloaded. Maybe shutting down kernels is not something we'd want to do here, but instead in [`on_shutdown`](https://www.mkdocs.org/dev-guide/plugins/#on_shutdown)? Clearing sessions should definitely happen in `on_post_build` though, not sure if that's only what `_shutdown_kernels` does :slightly_smiling_face:

pawamoy reviewed (changes requested):
LGTM, mostly style/lint suggestions :slightly_smiling_face: 

I didn't read the actual Jupyter/TypeScript-related code in details: I'll trust your evaluations in your own projects, and we can always fix things along the way.

benjamincburns commented:
Sorry for the radio silence on my end. Just to catch you up, we've been sprinting on our docs project pretty hard, and I haven't had time to come back to this review. A bunch of the changes to `markdown-exec` that we're using for our docs build are rather unique to our docs build process, and would probably be difficult to generalize at this stage.

Specifically to TS, I've switched over to using `tslab` instead of `deno`, and as a result I've disabled the type checking step for now, as that depended on `deno`. I think in retrospect that was a better fit for the `validator` phase, anyway. IIRC there weren't any changes required in `jupyter.py` to make this work, as I purposefully kept it quite language & kernel agnostic, however there were a few minor changes in `typescript.py` that were required, as I was passing some deno-specific kernel args.

Less specific to TS, we're converting a big pile of jupyter notebooks over to executable markdown, so for consistency with the old site we also modified the python formatter to use jupyter as well. I also went ahead and implemented the pre/post session hooks feature that I described in #73. The implementation for that is a bit janky, but if you're interested in bringing it back upstream I can reach out on a messaging app of choice and walk you through it.

pawamoy commented:
No worries, I wouldn't even consider that a radio silence :smile: 

> deno -> tslab

Fine by me, you're the expert.

> we also modified the python formatter to use jupyter as well

IIUC you're using a fork of Markdown Exec, right? At first we wouldn't apply this change upstream, as it's much heavier than just executing Python in the same process, and could lead to issues for current users. But I'm all for allowing users to choose one or the other, and for reconsidering later whether to switch defaults.

benjamincburns commented:
I'm sorry to say that we've abandoned this project for now. We weren't able to get things working well enough in the time that we'd budgeted, so we felt it would be best to call it a sunk cost and move on. As a result, my incentive to help maintain this is gone, so I feel the most responsible thing to do would be to close this.

I also don't think that this code should be merged into this project as-is. It's close, but I think I took a wrong turn by using Deno instead of tslab. In my testing Deno was too unstable, and had a tendency to crash too often due to unchecked errors.

I also think that type checking should be optional, and handled by the official typescript compiler. It should also have an option for the user to provide a path to their project's tsconfig.json (or one that's purpose-built for docs).

Finally, if anyone really wants TS support in markdown-exec and wants to pick up where I left off, I'd be happy to have a chat to walk you through what I've done here. My email is my GH username at gmail dot com.
